toke.c

```diff
@@ -4470,76 +4482,112 @@ S_intuit_more(pTHX_ char *s, char *e)
+                if (   len > 1
+                    && gv_fetchpvn_flags(tmpbuf,
+                                         len,
+                                         UTF ? SVf_UTF8 : 0,
+                                         SVt_PV))
                     weight -= 100;
-                else
+                else    /* Not a multi-char identifier already known in the
+                           program; is somewhat likely to be a subscript */
                     weight -= 10;
             }
-            else if (*s == '$'
+            else if (   *s == '$'
                      && s[1]
-                     && memCHRs("[#!%*<>()-=",s[1]))
+                     && memCHRs("[#!%*<>()-=", s[1]))
             {
-                if (/*{*/ memCHRs("])} =",s[2]))
+                /* Here we have what could be a punctuation variable.  If the
+                 * next character after it is a closing bracket, it makes it
+                 * quite likely to be that, and hence a subscript.  If it is
+                 * something else, more mildly a subscript */
+                if (/*{*/ memCHRs("])} =", s[2]))
                     weight -= 10;
                 else
                     weight -= 1;
             }
             break;
+
           case '\\':
             un_char = 254;
             if (s[1]) {
-                if (memCHRs("wds]",s[1]))
-                    weight += 100;
+                if (memCHRs("wds]", s[1]))
+                    weight += 100;  /* \w \d \s => strongly charclass */
                 else if (seen[(U8)'\''] || seen[(U8)'"'])
-                    weight += 1;
-                else if (memCHRs("rnftbxcav",s[1]))
-                    weight += 40;
+                    weight += 1;    /* \' => mildly charclass */
+                else if (memCHRs("rnftbxcav", s[1]))
+                    weight += 40;   /* \n, etc => charclass */
                 else if (isDIGIT(s[1])) {
-                    weight += 40;
+                    weight += 40;   /* \123 => charclass */
                     while (s[1] && isDIGIT(s[1]))
                         s++;
                 }
             }
-            else
+            else /* \ followed by NUL strongly indicates character class */
                 weight += 100;
             break;
+
           case '-':
+
+            /* If it is something like '-\', it is more likely to be a
+             * character class */
             if (s[1] == '\\')
                 weight += 50;
-            if (memCHRs("aA01! ",last_un_char))
+
+            /* If it is something like 'a-' or '0-', it is more likely to
+             * be a character class */
+            if (memCHRs("aA01! ", last_un_char))
                 weight += 30;
-            if (memCHRs("zZ79~",s[1]))
+
+            /* If it is something like '-Z' or '-7' (for octal) or '-9' it
+             * is more likely to be a character class */
+            if (memCHRs("zZ79~", s[1]))
```

Comment:
These comments really should mention the least obvious one of these: `[!-~]` (all printable ASCII characters).

Reply:
Thank you, I failed to spot this